File: js/problem61.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */

const getListLength = function(head) {
    if (head === null) {
        return 0;
    }
    let current = head, res = 0;
    while (current) {
        current = current.next;
        res++;
    }
    
    return res;
};

var rotateRight = function(head, k) {
    if ((!head) || (!head.next)) {
        return head;
    }
    let listLength = getListLength(head);
    k = k % listLength;
    if ((listLength === k) || (k === 0)) {
        return head;
    }
    
    let count = 1, runner = head;
    while (count < k) {
        runner = runner.next;
        count++;
    }
    
    let current = head, prev = null;
    while (runner.next) {
        prev = current;
        current = current.next;
        runner = runner.next;
    }
    runner.next = head;
    prev.next = null;
    
    return current;
};

File: leetcodeJava/src/main/java/Problem61.java
// 61. Rotate List

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    private int getListLength(ListNode head) {
        int res = 0;
        while (head != null) {
            res++;
            head = head.next;
        }
        return res;
    }
    
    public ListNode rotateRight(ListNode head, int k) {
        if ((head == null) || (head.next == null)) {
            return head;
        }
        int listLength = this.getListLength(head);
        k = k % listLength;
        if (k == 0) {
            return head;
        }
        ListNode runner = head;
        int count = 1;
        while (count < k) {
            count++;
            runner = runner.next;
        }
        ListNode current = head;
        ListNode previous = null;
        while (runner.next != null) {
            previous = current;
            current = current.next;
            runner = runner.next;
        }
        previous.next = null;
        runner.next = head;
        return current;
    }
}